### PR TITLE
Support threading in incompressible Navier Stokes

### LINKED ISSUE #15713: Support threading in incompressible Navier-Stokes
## Reason
The Griffin team is doing some MSR benchmarks which involves incompressible Navier-Stokes (INS) and they run a threading target. [This failed](https://moosebuild.inl.gov/job/275404/) because the `INSADObjectTracker` is a non-threaded object and we were setting a pointer to a function from multiple thread copies of a kernel. These pointers did not point to the same object because each thread has its own functions. This produced an error in `INSADObjectTracker` about consistency.

## Design
Force the `INSADObjectTracker` to use names instead of pointers to threaded data.

## Impact
Enable the Navier-Stokes module to be run with threads

### PR DESCRIPTION
**Pull request by lindsayad**
Force `INSADObjectTracker` to take names instead of pointers to threaded data.

Closes #15713

### CODE REVIEW

**Review by rwcarlsen — DISMISSED** [review]
Looks fine to me.  I'll give @YaqiWang a chance to chime in before merging.

File: framework/include/materials/MaterialProperty.h
Comment (YaqiWang) [inline framework/include/materials/MaterialProperty.h:185]: typically `name()` suggested a getter but this is a setter, so suggest to rename it to `setName`.

File: modules/navier_stokes/src/kernels/INSADBoussinesqBodyForce.C
Comment (YaqiWang) [inline modules/navier_stokes/src/kernels/INSADBoussinesqBodyForce.C:50]: `coupledComponents` can do the same as `getParam<std::vector<VariableName>>("temperature").size()`.

File: modules/navier_stokes/src/kernels/INSADEnergySource.C
Comment (YaqiWang) [inline modules/navier_stokes/src/kernels/INSADEnergySource.C:53]: same here.

File: modules/navier_stokes/src/kernels/INSADEnergySource.C
Comment (YaqiWang) [inline modules/navier_stokes/src/kernels/INSADEnergySource.C:55]: same here.

File: modules/navier_stokes/src/kernels/INSADBoussinesqBodyForce.C
Comment (YaqiWang) [inline modules/navier_stokes/src/kernels/INSADBoussinesqBodyForce.C:53]: `getVar("source_variable", 0)->name()` looks better.

File: modules/navier_stokes/src/kernels/INSADMomentumCoupledForce.C
Comment (YaqiWang) [inline modules/navier_stokes/src/kernels/INSADMomentumCoupledForce.C:52]: Not going to list all.

File: modules/navier_stokes/test/tests/ins/boussinesq/tests
Comment (YaqiWang) [inline modules/navier_stokes/test/tests/ins/boussinesq/tests:15]: I guess this is a short term solution before we add the multi-thread test target.

File: modules/navier_stokes/src/userobjects/INSADObjectTracker.C
Comment (YaqiWang) [inline modules/navier_stokes/src/userobjects/INSADObjectTracker.C:83]: Is it necessary to create this function? The diff indicates that we only need to change one place, which we can just directly use `compareCurrent`.

**Review by YaqiWang — CHANGES_REQUESTED** [review]
All comments are minor. One question is that will we add the multi-thread target? If so, one commit will need to be reverted in the future.

File: modules/navier_stokes/src/kernels/INSADBoussinesqBodyForce.C
Comment (YaqiWang) [reply modules/navier_stokes/src/kernels/INSADBoussinesqBodyForce.C:50]: I feel it is ok to change this to a `mooseError`.

File: modules/navier_stokes/src/kernels/INSADMomentumCoupledForce.C
Comment (lindsayad) [reply modules/navier_stokes/src/kernels/INSADMomentumCoupledForce.C:52]: I feel like this one is fine right? I don't see any `Coupleable` APIs that return the whole vector of coupled variables given a name

File: modules/navier_stokes/test/tests/ins/boussinesq/tests
Comment (lindsayad) [reply modules/navier_stokes/test/tests/ins/boussinesq/tests:15]: Yes although it would obviously work still even after we added the threading target. We would just have duplicate tests when running that target.

File: modules/navier_stokes/src/userobjects/INSADObjectTracker.C
Comment (lindsayad) [reply modules/navier_stokes/src/userobjects/INSADObjectTracker.C:83]: If I don't have this function then I have to do a whole specialization for `INSADObjectTracker::set` for `MooseEnum`, which is fine if that's what you prefer. Just more duplicate code (although only 10 lines)

**Review by YaqiWang — DISMISSED** [review]

**Comment by lindsayad** [comment]
Ok, I need a re-approval please :smile:

**Review by rwcarlsen — APPROVED** [review]

**Pull request merged**
